kafkaserver/handler.go

```diff
@@ -582,39 +579,66 @@ func (n *NewHandler) HandleSyncGroupRequest(_ *protocol.RequestHeader, req *prot
 	var resp protocol.ApiVersionsResponse
 	resp.ApiKeys = protocol.SupportedAPIVersions
 	return completionFunc(&resp)
 }
 
-func (n *NewHandler) SaslHandshakeRequestErrorResponse(errorCode int16, _ string, req *protocol.SaslHandshakeRequest) *protocol.SaslHandshakeResponse {
+func (n *connection) SaslHandshakeRequestErrorResponse(errorCode int16, _ string, req *protocol.SaslHandshakeRequest) *protocol.SaslHandshakeResponse {
 	var resp protocol.SaslHandshakeResponse
 	resp.ErrorCode = errorCode
 	return &resp
 }
 
-func (n *NewHandler) HandleSaslHandshakeRequest(_ *protocol.RequestHeader, _ *protocol.SaslHandshakeRequest, completionFunc func(resp *protocol.SaslHandshakeResponse) error) error {
+func (n *connection) HandleSaslHandshakeRequest(_ *protocol.RequestHeader, req *protocol.SaslHandshakeRequest, completionFunc func(resp *protocol.SaslHandshakeResponse) error) error {
 	var resp protocol.SaslHandshakeResponse
-	// For now, we don't implement any mechanisms, this is just a placeholder
+	conversation, ok, err := n.s.saslAuthManager.CreateConversation(*req.Mechanism)
+	if err != nil {
+		return err
+	}
+	if !ok {
+		resp.ErrorCode = protocol.ErrorCodeUnsupportedSaslMechanism
+	} else {
+		n.saslConversation = conversation
+	}
+	saslScram256 := auth.AuthenticationSaslScramSha256
+	saslScram512 := auth.AuthenticationSaslScramSha512
+	resp.Mechanisms = []*string{&saslScram256, &saslScram512}
 	return completionFunc(&resp)
 }
 
-func (n *NewHandler) SaslAuthenticateRequestErrorResponse(errorCode int16, _ string, req *protocol.SaslAuthenticateRequest) *protocol.SaslAuthenticateResponse {
+func (n *connection) SaslAuthenticateRequestErrorResponse(errorCode int16, _ string, req *protocol.SaslAuthenticateRequest) *protocol.SaslAuthenticateResponse {
 	var resp protocol.SaslAuthenticateResponse
 	resp.ErrorCode = errorCode
 	return &resp
 }
 
-func (n *NewHandler) HandleSaslAuthenticateRequest(_ *protocol.RequestHeader, req *protocol.SaslAuthenticateRequest, completionFunc func(resp *protocol.SaslAuthenticateResponse) error) error {
+func (n *connection) HandleSaslAuthenticateRequest(_ *protocol.RequestHeader, req *protocol.SaslAuthenticateRequest, completionFunc func(resp *protocol.SaslAuthenticateResponse) error) error {
 	var resp protocol.SaslAuthenticateResponse
-	// Always fail, just a placeholder
-	resp.ErrorCode = protocol.ErrorCodeSaslAuthenticationFailed
+	conv := n.saslConversation
+	if conv == nil {
+		resp.ErrorCode = protocol.ErrorCodeIllegalSaslState
+		msg := "SaslAuthenticateRequest without a preceding SaslAuthenticateRequest"
+		resp.ErrorMessage = &msg
+	} else {
+		saslRespBytes, complete, failed := conv.Process(req.AuthBytes)
+		if failed {
+			resp.ErrorCode = protocol.ErrorCodeSaslAuthenticationFailed
+		} else {
+			resp.AuthBytes = saslRespBytes
+			if complete {
+				principal := conv.Principal()
+				n.authContext.Principal = &principal
+				n.authContext.Authenticated = true
```

Comment:
IIUC, technically `complete` does not equal authenticated, as Process returns `Done()` not `Valid()`.

Reply:
as above